ngraph/core/src/op/if.cpp

```diff
@@ -0,0 +1,292 @@
+        }
+    }
+}
+
+void op::v8::If::fill_body(const std::shared_ptr<op::v8::If>& new_op,
+                           size_t branch_index,
+                           const OutputVector& new_args) const
+{
+    auto body = m_bodies[branch_index];
+    auto param_size = body->get_parameters().size();
+    std::vector<::ngraph::element::Type> types(param_size);
+    std::vector<::ngraph::PartialShape> new_shapes(param_size);
+    auto& input_descriptions = m_input_descriptions[branch_index];
+    size_t parameters_num = 0;
+    for (auto& input_description : input_descriptions)
+    {
+        if (input_description->m_input_index < new_args.size())
+        {
+            types[input_description->m_body_parameter_index] =
+                new_args[input_description->m_input_index].get_element_type();
+            new_shapes[input_description->m_body_parameter_index] =
+                new_args[input_description->m_input_index].get_partial_shape();
+            parameters_num++;
+        }
+    }
+    auto func =
+        std::make_shared<Function>(body->get_results(), body->get_sinks(), body->get_parameters());
+    auto spec_func =
+        specialize_function(func, types, new_shapes, std::vector<void*>(parameters_num, nullptr));
+    new_op->m_bodies[branch_index] = std::make_shared<Function>(
+        spec_func->get_results(), spec_func->get_sinks(), spec_func->get_parameters());
+
+    for (const auto& input_description : input_descriptions)
+    {
+        new_op->m_input_descriptions[branch_index].push_back(input_description->copy());
+    }
+    for (const auto& output_description : m_output_descriptions[branch_index])
+    {
+        new_op->m_output_descriptions[branch_index].push_back(output_description->copy());
+    }
+}
+
+std::shared_ptr<Node> op::v8::If::clone_with_new_inputs(const OutputVector& new_args) const
+{
+    NGRAPH_OP_SCOPE(v8_If_clone_with_new_inputs);
+    auto op = make_shared<op::v8::If>(new_args);
+    NGRAPH_CHECK(op.get(),
+                 op != nullptr,
+                 "Cannot clone ",
+                 description(),
+                 " operation with name ",
+                 get_friendly_name());
+    op->set_output_size(m_output_descriptions[0].size());
+    fill_body(op, then_body_index, new_args);
+    fill_body(op, else_body_index, new_args);
+    op->validate_and_infer_types();
+    return op;
+}
+
+bool op::v8::If::evaluate(const HostTensorVector& outputs, const HostTensorVector& inputs) const
+{
+    NGRAPH_OP_SCOPE(v8_If_evaluate);
+    runtime::reference::if_reference(
+        m_bodies, m_output_descriptions, m_input_descriptions, outputs, inputs);
+    return true;
+}
+
+void op::v8::If::set_input(const Output<Node>& value,
+                           const std::shared_ptr<Parameter>& then_parameter,
+                           const std::shared_ptr<Parameter>& else_parameter)
+{
+    ParameterVector param_vec;
+    if (then_parameter != nullptr)
```

Comment:
why do we need this check?

Reply:
It is necessary to aviod passing nullptr to MultiSubgraphOp::set_invariant_input

Reply:
I don't think this is necessary, nothing seems to happen if the parameters are nullptr here
![image](https://user-images.githubusercontent.com/21181928/127447109-1f85e898-eff6-47de-bff1-c717dcaffcca.png)

Reply:
it's better to add NGRAPH_CHECK(..) to avoid then_parameter or else_parameter are nullptrs

Reply:
set_input and set_output were rewrited.